GH-33592: [C++] support casting nullable fields to non-nullable if there are no null values

NickCrews opened the pull request:
* GitHub Issue: #33592

Notes for myself/fixer:
- [tests that need to get updated](https://github.com/search?q=repo%3Aapache%2Farrow%20%22cannot%20cast%20nullable%20field%22&type=code) (almost definitely not a complete list)
- [update: actually we should handle the go implementation in the go repository.] hmm, looks like [go wrapper does its own nullability checks](https://github.com/apache/arrow/blob/f078942ce2df68de8f48c3b4233132133601ca53/go/arrow/compute/cast.go#L283-L286). I assume this is just an optimization to not have to go into the C++ and hit the error down there. So if we just delete this check then I think the C++ logic will handle all of it???
- [how the plain column handles this cast](https://github.com/apache/arrow/blob/f078942ce2df68de8f48c3b4233132133601ca53/python/pyarrow/table.pxi#L3238-L3243), some logic like this probably needs to get ported over to the struct implementation
- (running from /cpp/build) `cmake .. --preset ninja-debug-basic`, then `cmake --build . && PYTHON=python ctest -R 'arrow-compute-scalar-cast-test' --output-on-failure` to run the specific test
-  to format: `uvx pre-commit run --all-files clang-format`

NickCrews commented:
Thanks for the help! I'm a random contributor, so I'm not sure about the exact workflow, but I think both getting the CI run approved and having a C++ owner approve it are needed.

Any more detailed thoughts on if the tests are adequate, if there's anywhere else in the code base that you think needs to change, my handling of the go implementation by just deleting the shortcircuit, or any other more detailed thoughts? Basically anything that would reduce the mental load on the code owner I think would increase the odds they approve it :)

mapleFU commented:
You can mark as ready for review? Or this is still wip?

NickCrews commented:
Oops, didn't mean for this to still be marked WIP. Looks like a few failures that need to get fixed, but still a high-level review of the general approach would still be appreciated in the meantime.

zeroshade commented:
You probably need the same corresponding check on the Go side to verify that it's only allowed if there are no nulls.

Also, the Go implementation has been moved to the apache/arrow-go repository. So please file the PR there for the Go side. Sorry for the confusion, we just haven't removed the Go code from here yet.

NickCrews commented:
Thanks @zeroshade , will do. Do we need both PRs to land at about the same time, or can I do that independently?

I will undo my changes to the go code here, leaving it untouched.

zeroshade commented:
They can land independently, no issues there. Thanks!

NickCrews commented:
Just pushed a new version:
- dropped the Go changes
- rebased on top of 3600db8c74f2d83850ad510aa7efbe10fddd9600, which is one commit behind [main](https://github.com/apache/arrow/commits/main/), because `main` is failing CI, but the commit I chose passed CI. I did this because I think the failing CI checks in this PR are not related to this PR
- I discovered `archery` and formatted the files with `archery lint --clang-format --fix`

We will see if this passes CI now...

NickCrews commented:
Pushed a new version, hopefully that fixed the broken tests:
- To go from nullable to non-nullable type, need to use the unsafe cast option
- fix typo of int8 to int64 so the precision matches
- removed the [Go] tag from the commit message

NickCrews commented:
@zeroshade I think this is ready to review/merge, the failing CI runs look like flakes when trying to setup the environment?

kou commented:
Can we move this to apache/arrow-go?

kou commented:
Ah, the Go part was removed from this PR.
I've removed "[Go]" from the PR title.

NickCrews commented:
Those CI failures look unrelated? Not sure, [CI history looks pretty flaky](https://github.com/apache/arrow/commits/main/) to me.

NickCrews commented:
The current CI failures look to be caused by https://github.com/apache/arrow/issues/44985 (or similar), not by this PR.

I rebased on main, everything looks pretty much the same as before.

@zeroshade can you give this a review when you get the chance, or let me know how I can make it easier for you to do so? @rustyconover if you want to take a look again, maybe that will help. Thank you both!

zanmato1984 commented on cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:
```diff
@@ -404,14 +400,20 @@ struct CastStruct {
       } else {
         const auto& values = (in_array.child_data[field_index].ToArrayData()->Slice(
             in_array.offset, in_array.length));
+        if (!target_field->nullable() && values->null_count > 0) {
```
Is it possible if `values->null_count` is `kUnknownNullCount`? In which case, we should probably use child array's `GetNullCount()` instead?

And if we are not sure about that, can we have a dedicated test for it?

mapleFU replied:
I think `values->null_count` can likely to be `kUnknownNullCount`

NickCrews replied:
I have never heard of kUnknownNullCount. I'm not sure what we should do here, I think I need to know a lot more background. Can someone who has experience with null counts help?

Some questions this brings up:
1. This sounds like we sometimes don't know the null count, I assume because it is expensive to compute. Would this be a good reason to actually force computation to happen (behavior A), or should we be conservative, and if we find kUnkownNullCount, then we assume that there could be nulls and don't let the cast happen and therefore error (behavior B). Behavior A seems much more correct to me.
2. I don't think this function should have the responsibility of looking at the child arrays null counts. Should we add some higher level API to arrays like nullCountOrChildNullCount() that actually does this messy work??

zanmato1984 replied:
> I have never heard of kUnknownNullCount. I'm not sure what we should do here, I think I need to know a lot more background. Can someone who has experience with null counts help?

Sorry I'm not sure if there is a well-documented rule of dealing with `kUnknownNullCount`, maybe @pitrou does? But in practice, I (and many other developers I suppose) just keep in mind that the null count can be unknown and is lazily computed. One can enforce such computation once the null count is absolutely needed (there are various APIs helping on this).
 
> 1. This sounds like we sometimes don't know the null count, I assume because it is expensive to compute. Would this be a good reason to actually force computation to happen (behavior A), or should we be conservative, and if we find kUnkownNullCount, then we assume that there could be nulls and don't let the cast happen and therefore error (behavior B). Behavior A seems much more correct to me.

Yeah your assumption is right. And I would go A (enforcing the null count computation) because IMO this is essential for what this PR is trying to do. Plus, null count is unknown by default, so B will make this change almost useless.

> 2. I don't think this function should have the responsibility of looking at the child arrays null counts. Should we add some higher level API to arrays like nullCountOrChildNullCount() that actually does this messy work??

IMO this is where you must make contact with child arrays, given that you are in the context of dealing with a `Struct` array - which is essentially a container of child arrays. And the original code already does so. Defining an API like `nullCountOrChildNullCount()` could be semantically hard/useless because: 1) the validity (null or not) of the whole struct and 2) the validity of each individual field (and you have to specify which field), mean very different things. So the null count of a particular field is a property of this particular field only and should be computed/checked individually.

pitrou replied:
You should just call `GetNullCount` so that the null count is computed if not already known.

NickCrews replied:
OK, I adjusted to `values->GetNullCount() > 0`, I think we all the issues brought up here are resolved. re-open if I missed something.

zanmato1984 reviewed (commented):
A small question about null count checking.

mapleFU commented on cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:
```diff
@@ -404,14 +400,20 @@ struct CastStruct {
         const auto& values = (in_array.child_data[field_index].ToArrayData()->Slice(
             in_array.offset, in_array.length));
+        if (!target_field->nullable() && values->null_count > 0) {
+          return Status::TypeError("field '", target_field->name(),
```
Besides, should we discard the `buffer[0]` if new data doesn't contains any nulls?

NickCrews replied:
@mapleFU can you elaborate a little more? I don't think I understand neither why we should do this, or how. Does buffer[0] contain the null bitmap?

mapleFU replied:
Ignore me, i previously think we can

```
        auto values = (in_array.child_data[field_index].ToArrayData()->Slice(
            in_array.offset, in_array.length));
       if (out field not nullable but in field nullable) {
          if (values->GetNullCount() > 0) {
             return error
          }
          values->buffers[0].reset();
       }
```

But I also think this doesn't matter here

pitrou commented on cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:
```diff
@@ -404,14 +400,20 @@ struct CastStruct {
         const auto& values = (in_array.child_data[field_index].ToArrayData()->Slice(
             in_array.offset, in_array.length));
+        if (!target_field->nullable() && values->null_count > 0) {
+          return Status::TypeError("field '", target_field->name(),
```
Since this is data-dependent, this should be `Status::Invalid`.

NickCrews replied:
Fixed this, thank you1

NickCrews commented:
@pitrou @mapleFU I applied your two suggestions, and I think this is ready for review again. Still one outstanding question about `buffer[0]`. CI is passing, except for one flaky test that failed during git checkout.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_cast_test.cc:
```diff
@@ -2964,28 +2964,31 @@ TEST(Cast, StructToDifferentNullabilityStruct) {
         std::make_shared<Field>("a", int64(), false),
         std::make_shared<Field>("c", int64(), false)};
     const auto dest2_non_nullable = arrow::struct_(fields_dest2_non_nullable);
-    const auto options2_non_nullable = CastOptions::Safe(dest2_non_nullable);
+    const auto options2_non_nullable = CastOptions::Unsafe(dest2_non_nullable);
```
Why do we need to switch from `Safe` to `Unsafe` here?

NickCrews replied:
Thanks, you're right, I'm not sure why I had this change, maybe it wasn't passing in some earlier version of this changeset unless I had it. But it now works just fine with Safe, so reverting to that.

pitrou commented on cpp/src/arrow/compute/kernels/scalar_cast_test.cc:
```diff
@@ -2964,28 +2964,31 @@ TEST(Cast, StructToDifferentNullabilityStruct) {
+    ASSERT_OK_AND_ASSIGN(auto dest_no_nulls,
+                         StructArray::Make({c_dest_no_nulls}, fields_dest_no_nulls));
+    const auto options3_non_nullable =
+        CastOptions::Unsafe(arrow::struct_(fields_dest_no_nulls));
```
Similarly, why `Unsafe`?

NickCrews replied:
Thanks, you are right.

mapleFU commented on cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:
```diff
@@ -404,14 +400,20 @@ struct CastStruct {
         const auto& values = (in_array.child_data[field_index].ToArrayData()->Slice(
             in_array.offset, in_array.length));
+        if (!target_field->nullable() && values->GetNullCount() > 0) {
+          return Status::Invalid("field '", target_field->name(),
```
Also `in_field->name()` and `in_field->type->ToString()`?

NickCrews replied:
Good idea, adding that.

mapleFU commented on cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:
```diff
@@ -404,14 +400,20 @@ struct CastStruct {
       } else {
         const auto& values = (in_array.child_data[field_index].ToArrayData()->Slice(
             in_array.offset, in_array.length));
+        if (!target_field->nullable() && values->GetNullCount() > 0) {
```
should we check `in_field->nullable()`?

NickCrews replied:
For correctness, I don't think this is needed (because the `values->GetNullCount() > 0` will take care of this), but for performance I could see it being nice to skip the `values->GetNullCount()` computation. I'll add the check.

mapleFU replied:
Yeah this might cause the null bitmap being counted, I think check nullable would be faster than checking this if not neccessary

mapleFU commented on cpp/src/arrow/compute/kernels/scalar_cast_test.cc:
```diff
@@ -2974,18 +2975,20 @@ TEST(Cast, StructToDifferentNullabilityStruct) {
 
-    std::vector<std::shared_ptr<Field>> fields_dest3_non_nullable = {
+    std::shared_ptr<Array> c_dest_no_nulls;
+    c_dest_no_nulls = ArrayFromJSON(int64(), "[9, 11, 44]");
```
Can we add a test for sub slices? Like: `[null, 9, 11, 14]`, the `slice(0,2)` would cast failed, and `[11, 14]` would success?

NickCrews replied:
that seems like a good idea. Adding it now.

NickCrews replied:
Added them.

mapleFU reviewed (commented):
General LGTM!

NickCrews commented:
whoops, I accidentally force-pushed with no changes, which closed the PR. Now it's back up

NickCrews commented:
@mapleFU as I added the slicing tests, I also refactored the tests to make them much more consistent and concise. Take another look and make sure that your "LGTM" still holds after those changes.

NickCrews commented on cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:
```diff
@@ -403,17 +399,27 @@ struct CastStruct {
+          return Status::Invalid("field '", in_field->name(), "' of type ",
+                                 in_field_type->ToString(),
+                                 " has nulls. Can't cast to field '", out_field->name(),
+                                 "' of non-nullable type ", out_field_type->ToString());
```
good catch, thank you! Fixed now.

NickCrews commented on cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:
```diff
@@ -403,17 +399,27 @@ struct CastStruct {
             in_array.offset, in_array.length));
-        ARROW_ASSIGN_OR_RAISE(Datum cast_values,
-                              Cast(values, target_type, options, ctx->exec_context()));
+        if (in_field->nullable() && !out_field->nullable() &&
```
should we put an UNLIKELY compiler hint here?

pitrou replied:
It's not really important, this is called only once per field (but if you want to, there's a `ARROW_PREDICT_FALSE` macro that you can use).

NickCrews replied:
I'll keep it simple and leave it out.

NickCrews commented:
I'm getting these failing UBSAN errors in CI. See below for the relevant logs. Is it because of the new `std::vector<std::shared_ptr<Array>> arrays_dest_ac = {arrays_dest[0], arrays_dest[2]}` code that I added? Is there a good way to silence/ignore this error?

<details>
<summary>Details</summary>

```
[ RUN      ] Cast.StructToStructSubset
/usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:1064:34: runtime error: addition of unsigned offset to 0x60700006f300 overflowed to 0x60700006f2e0
    #0 0x7f811aca4bc8 in std::vector<std::shared_ptr<arrow::Field>, std::allocator<std::shared_ptr<arrow::Field> > >::operator[](unsigned long) const /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:1064:34
    #1 0x7f811ac9b27a in arrow::DataType::field(int) const /arrow/cpp/src/arrow/type.h:153:61
    #2 0x7f8105e4f8f1 in arrow::compute::internal::(anonymous namespace)::CastStruct::Exec(arrow::compute::KernelContext*, arrow::compute::ExecSpan const&, arrow::compute::ExecResult*) /arrow/cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:403:38
    #3 0x7f8105ab206e in arrow::compute::detail::(anonymous namespace)::ScalarExecutor::ExecuteNonSpans(arrow::compute::detail::ExecListener*) /arrow/cpp/src/arrow/compute/exec.cc:920:7
    #4 0x7f8105aa99fa in arrow::compute::detail::(anonymous namespace)::ScalarExecutor::Execute(arrow::compute::ExecBatch const&, arrow::compute::detail::ExecListener*) /arrow/cpp/src/arrow/compute/exec.cc:810:14
    #5 0x7f8105c471a6 in arrow::compute::detail::FunctionExecutorImpl::Execute(std::vector<arrow::Datum, std::allocator<arrow::Datum> > const&, long) /arrow/cpp/src/arrow/compute/function.cc:278:5
    #6 0x7f8105c17bfe in arrow::compute::(anonymous namespace)::ExecuteInternal(arrow::compute::Function const&, std::vector<arrow::Datum, std::allocator<arrow::Datum> >, long, arrow::compute::FunctionOptions const*, arrow::compute::ExecContext*) /arrow/cpp/src/arrow/compute/function.cc:343:21
    #7 0x7f8105c16d84 in arrow::compute::Function::Execute(std::vector<arrow::Datum, std::allocator<arrow::Datum> > const&, arrow::compute::FunctionOptions const*, arrow::compute::ExecContext*) const /arrow/cpp/src/arrow/compute/function.cc:350:10
    #8 0x7f8105a5d13c in arrow::compute::internal::(anonymous namespace)::CastMetaFunction::ExecuteImpl(std::vector<arrow::Datum, std::allocator<arrow::Datum> > const&, arrow::compute::FunctionOptions const*, arrow::compute::ExecContext*) const /arrow/cpp/src/arrow/compute/cast.cc:124:23
    #9 0x7f8105c21b5c in arrow::compute::MetaFunction::Execute(std::vector<arrow::Datum, std::allocator<arrow::Datum> > const&, arrow::compute::FunctionOptions const*, arrow::compute::ExecContext*) const /arrow/cpp/src/arrow/compute/function.cc:483:10
    #10 0x7f8105a8cfc8 in arrow::compute::CallFunction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<arrow::Datum, std::allocator<arrow::Datum> > const&, arrow::compute::FunctionOptions const*, arrow::compute::ExecContext*) /arrow/cpp/src/arrow/compute/exec.cc:1369:16
    #11 0x55f68eeb23c9 in arrow::compute::CheckScalarNonRecursive(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::vector<arrow::Datum, std::allocator<arrow::Datum> > const&, arrow::Datum const&, arrow::compute::FunctionOptions const*) /arrow/cpp/src/arrow/compute/kernels/test_util_internal.cc:80:3
    #12 0x55f68eeb8480 in arrow::compute::CheckScalar(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::vector<arrow::Datum, std::allocator<arrow::Datum> > const&, arrow::Datum, arrow::compute::FunctionOptions const*) /arrow/cpp/src/arrow/compute/kernels/test_util_internal.cc:109:3
    #13 0x55f68eecc36c in arrow::compute::CheckScalarUnary(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, arrow::Datum, arrow::Datum, arrow::compute::FunctionOptions const*) /arrow/cpp/src/arrow/compute/kernels/test_util_internal.cc:255:3
    #14 0x55f68ef4d8f7 in arrow::compute::CheckCast(std::shared_ptr<arrow::Array>, std::shared_ptr<arrow::Array>, arrow::compute::CastOptions) /arrow/cpp/src/arrow/compute/kernels/scalar_cast_test.cc:111:3
    #15 0x55f68f079ab8 in arrow::compute::CheckStructToStructSubset(std::vector<std::shared_ptr<arrow::DataType>, std::allocator<std::shared_ptr<arrow::DataType> > > const&) /arrow/cpp/src/arrow/compute/kernels/scalar_cast_test.cc:3829:7
    #16 0x55f68f06d09f in arrow::compute::Cast_StructToStructSubset_Test::TestBody() /arrow/cpp/src/arrow/compute/kernels/scalar_cast_test.cc:4003:36
    #17 0x7f811af8260e in void testing::internal::HandleExceptionsInMethodIfSupported<testing::Test, void>(testing::Test*, void (testing::Test::*)(), char const*) (/usr/local/lib/libarrow_testing.so.2000+0x108c60e) (BuildId: 408b230591f17f240ad9d2b7d868274f1179009d)
    #18 0x7f811af770c5 in testing::Test::Run() (/usr/local/lib/libarrow_testing.so.2000+0x10810c5) (BuildId: 408b230591f17f240ad9d2b7d868274f1179009d)
    #19 0x7f811af77244 in testing::TestInfo::Run() (/usr/local/lib/libarrow_testing.so.2000+0x1081244) (BuildId: 408b230591f17f240ad9d2b7d868274f1179009d)
    #20 0x7f811af777f8 in testing::TestSuite::Run() (/usr/local/lib/libarrow_testing.so.2000+0x10817f8) (BuildId: 408b230591f17f240ad9d2b7d868274f1179009d)
    #21 0x7f811af77efe in testing::internal::UnitTestImpl::RunAllTests() (/usr/local/lib/libarrow_testing.so.2000+0x1081efe) (BuildId: 408b230591f17f240ad9d2b7d868274f1179009d)
    #22 0x7f811af82bd6 in bool testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*) (/usr/local/lib/libarrow_testing.so.2000+0x108cbd6) (BuildId: 408b230591f17f240ad9d2b7d868274f1179009d)
    #23 0x7f811af7730b in testing::UnitTest::Run() (/usr/local/lib/libarrow_testing.so.2000+0x108130b) (BuildId: 408b230591f17f240ad9d2b7d868274f1179009d)
    #24 0x55f68edefb33 in main (/build/cpp/debug/arrow-compute-scalar-cast-test+0x2adb33) (BuildId: c176b331f0537175685c7888a83fc8e112006f6a)
    #0 0x7f80f7d06d8f in
    #26 0x7f80f7d06e3f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x29e3f) (BuildId: cd410b710f0f094c6832edd95931006d883af48e)
    #27 0x55f68edefb94 in _start (/build/cpp/debug/arrow-compute-scalar-cast-test+0x2adb94) (BuildId: c176b331f0537175685c7888a83fc8e112006f6a)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /usr/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/stl_vector.h:1064:34 in
/build/cpp/src/arrow/compute/kernels

      Start 22: arrow-compute-scalar-cast-test
    Test #22: arrow-compute-scalar-cast-test ...............***Failed   12.44 sec
```

</details>

pitrou commented:
Well, the Windows 2019 CI test shows a similar error, so I think it needs diagnosing and solving:
```
[ RUN      ] Cast.StructToStructSubset
C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\include\vector(1563) : Assertion failed: vector subscript out of range
```

NickCrews commented:
I rebased on top of https://github.com/apache/arrow/pull/45500 so that hopefully CI will pass

NickCrews commented on cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:
```diff
@@ -403,17 +399,27 @@ struct CastStruct {
-        ARROW_ASSIGN_OR_RAISE(auto nulls,
-                              MakeArrayOfNull(target_type->GetSharedPtr(), batch.length));
+    for (int in_field_index : fields_to_select) {
+      const auto& in_field = in_type.field(in_field_index);
```
Thank you, good eye, I bet that is the problem. Another example of why we owe a beer to everyone who works on sanitizers/fuzzers/etc 🙇

pitrou reviewed (approved):
+1. Thanks for this @NickCrews !

pitrou commented:
@github-actions crossbow submit -g cpp

pitrou commented:
The CI failures are unrelated (see https://github.com/apache/arrow/pull/45524).

NickCrews commented:
Thank you everyone! That was a lot of effort everyone did, I wouldn't have been able to do that without all the help. I'm excited for this to be released which will allow me to remove some hacky workarounds I have in my app code.
